app_dart/lib/src/request_handlers/task_status_subscription.dart

```diff
@@ -0,0 +1,119 @@
+// Copyright 2019 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:convert';
+
+import 'package:cocoon_service/src/model/luci/push_message.dart';
+import 'package:cocoon_service/src/request_handling/subscription_handler.dart';
+import 'package:gcloud/db.dart';
+import 'package:meta/meta.dart';
+
+import '../model/appengine/commit.dart';
+import '../model/appengine/task.dart';
+import '../request_handling/authentication.dart';
+import '../request_handling/body.dart';
+import '../request_handling/exceptions.dart';
+import '../service/cache_service.dart';
+import '../service/config.dart';
+import '../service/datastore.dart';
+import '../service/logging.dart';
+
+/// An endpoint for listening to build updates for task updates.
+///
+/// The PubSub subscription is set up here:
+/// https://cloud.google.com/cloudpubsub/subscription/detail/luci-builds?project=flutter-dashboard&tab=overview
+///
+/// This endpoint is responsible for updating Datastore with the result of tasks from LUCI.
+@immutable
+class TaskStatusSubscription extends SubscriptionHandler {
+  const TaskStatusSubscription(
+    CacheService cache,
+    Config config,
+    AuthenticationProvider authProvider, {
+    @visibleForTesting this.datastoreProvider = DatastoreService.defaultProvider,
+  }) : super(cache: cache, config: config, authProvider: authProvider, topicName: 'luci-builds');
+
+  final DatastoreServiceProvider datastoreProvider;
+
+  @override
+  Future<Body> post() async {
+    final String data = message.data!;
+    final BuildPushMessage buildPushMessage =
+        BuildPushMessage.fromJson(json.decode(String.fromCharCodes(base64.decode(data))) as Map<String, dynamic>);
+
+    final Map<String, dynamic> userData = jsonDecode(buildPushMessage.userData!) as Map<String, dynamic>;
+    final String? commitBranch = userData['commit_branch'] as String?;
+    if (commitBranch == null) {
+      throw const BadRequestException('userData does not contain commit_branch');
+    }
+    final String? commitSha = userData['commit_sha'] as String?;
+    if (commitSha == null) {
+      throw const BadRequestException('userData does not contain commit_sha');
+    }
+    final String? builderName = userData['builder_name'] as String?;
+    if (builderName == null) {
+      throw const BadRequestException('userData does not contain builder_name');
+    }
+    final String? newStatus = buildPushMessage.build?.status.toString();
+    if (newStatus == null) {
+      throw const BadRequestException('userData does not contain status');
+    }
+
+    final DatastoreService datastore = datastoreProvider(config.db);
+
+    // if (newStatus != Task.statusSucceeded && newStatus != Task.statusFailed) {
```

Comment:
@CaseyHillers This is failing, because the Task.statusSucceeded ('Succeeded') and Task.statusFailed ('Failed') does not match the push_message.status values (COMPLETED, SUCCESS, etc.). Do we need to translate this over to Task enums for it to work?